src/procedures/proc_pagerank.c

```diff
@@ -95,59 +95,52 @@ ProcedureResult Proc_PagerankInvoke(ProcedureCtx *ctx,
-	 * nodes of type 'l'
-	 *
-	 * if relation is specified:
-	 * cast 'r' to a boolean matrix */
-	if(label != NULL || relation != NULL) {
+	 * nodes of type 'l' */
+	if(label != NULL) {
 		//----------------------------------------------------------------------
 		// Create a NxN matrix, one row for each labeled entity
 		//----------------------------------------------------------------------
-		if(label) {
-			info = GrB_Matrix_nvals(&n, l);
-			ASSERT(info == GrB_SUCCESS);
-		}
+		info = GrB_Matrix_nvals(&n, l);
+		ASSERT(info == GrB_SUCCESS);
 
 		GrB_Matrix reduced; // Relation matrix reduced to only 'l' rows/cols
 		info = GrB_Matrix_new(&reduced, GrB_BOOL, n, n);
 		ASSERT(info == GrB_SUCCESS);
 
 		// Discard rows of 'r' associated with nodes of a different type than 'l'
 		// this will also perform casting to boolean.
-		if(n != nrows) {
-			mapping = rm_malloc(sizeof(GrB_Index) * n);
-			// Extract row indecies from 'l', coresponding to node IDs.
-			info = GrB_Matrix_extractTuples_BOOL(mapping, GrB_NULL, GrB_NULL, &n, l);
-			ASSERT(info == GrB_SUCCESS);
-
-			info = GrB_Matrix_extract(reduced, GrB_NULL, GrB_NULL, r, mapping, n,
-					mapping, n, GrB_NULL);
-			ASSERT(info == GrB_SUCCESS);
-		} else {
-			/* There no need to perform extraction as either 'l' isn't specified
-			 * if if 'l' is given, 'r' dimension is NxN the same as
-			 * the number of entries in 'l' which means all connections
-			 * described in 'r' connect nodes of type 'l'
-			 * Unfortunately we still need to type cast 'r' to boolean */
-			GrB_Descriptor desc;
-			GrB_Descriptor_new(&desc);
-			GrB_Descriptor_set(desc, GrB_INP0, GrB_TRAN);
-			info = GrB_transpose(reduced, GrB_NULL, GrB_NULL, r, desc);
-			ASSERT(info == GrB_SUCCESS);
-			GrB_free(&desc);
-		}
+		mapping = rm_malloc(sizeof(GrB_Index) * n);
+		// Extract row indecies from 'l', coresponding to node IDs.
+		info = GrB_Matrix_extractTuples_BOOL(mapping, GrB_NULL, GrB_NULL, &n, l);
+		ASSERT(info == GrB_SUCCESS);
+
+		info = GrB_Matrix_extract(reduced, GrB_NULL, GrB_NULL, r, mapping, n,
+								  mapping, n, GrB_NULL);
+		ASSERT(info == GrB_SUCCESS);
 
 		r = reduced;
-		free_r = true;
+	} else {
+		//----------------------------------------------------------------------
+		// Make a copy of the relationship matrix,
+		// truncating it to remove unused rows and casting it to boolean.
+		//----------------------------------------------------------------------
+		info = GrB_Matrix_nrows(&nrows, r);
+		ASSERT(info == GrB_SUCCESS);
+
+		// Build a new boolean matrix of the full dimensions.
+		GrB_Matrix r_trimmed;
+		info = GrB_Matrix_new(&r_trimmed, GrB_BOOL, nrows, nrows);
+		ASSERT(info == GrB_SUCCESS);
+		// Copy the relationship matrix, casting contents to boolean.
+		info = GrB_Matrix_dup(&r_trimmed, r);
+		ASSERT(info == GrB_SUCCESS);
+
+		// Resize to remove unused rows.
+		n = Graph_UncompactedNodeCount(g);
+		GxB_Matrix_resize(r_trimmed, n, n);
+		r = r_trimmed;
```

Comment:
you assign to `r` without freeing the old value

Reply:
`r` before here is an actual adjacency matrix, not a matrix allocated for this query.

Reply:
I think `GrB_Matrix_dup` allocates a new matrix
 in GraphBLAS GB_dup2.c
```c
GrB_Info GB_dup2            // make an exact copy of a matrix
(
    GrB_Matrix *Chandle,    // handle of output matrix to create 
    const GrB_Matrix A,     // input matrix to copy
    const bool numeric,     // if true, duplicate the numeric values
    const GrB_Type ctype,   // type of C, if numeric is false
    GB_Context Context
)
{

    //--------------------------------------------------------------------------
    // determine the number of threads to use
    //--------------------------------------------------------------------------

    GB_GET_NTHREADS_MAX (nthreads_max, chunk, Context) ;

    //--------------------------------------------------------------------------
    // C = A
    //--------------------------------------------------------------------------

    if (A->nvec_nonempty < 0)
    { 
        A->nvec_nonempty = GB_nvec_nonempty (A, Context) ;
    }

    // [ create C; allocate C->p and do not initialize it
    // C has the exact same hypersparsity as A.
    GrB_Info info ;
    int64_t anz = GB_NNZ (A) ;

    // allocate a new header for C if (*Chandle) is NULL, or reuse the
    // existing header if (*Chandle) is not NULL.
    GrB_Matrix C = (*Chandle) ;


    GB_CREATE (&C, numeric ? A->type : ctype, A->vlen, A->vdim, GB_Ap_malloc,
        A->is_csc, GB_SAME_HYPER_AS (A->is_hyper), A->hyper_ratio, A->plen,
        anz, true, Context) ;
```

Reply:
Oh, you're right! I misread the documentation on this function. Good catch! (Also, it worries me that my local Valgrind didn't catch this.)